### DISCUSSION TITLE
How to do Update multiple columns with WHERE EXISTS

### DISCUSSION THREAD

**Discussion by tb-mtg** [linq2db/linq2db#4697] [Q&A]
How do I perform the following SQL in linq2db which is valid in DB2?
```
UPDATE PROJECT P
SET (DEPTNO, EMPID) = (
  SELECT WORKDEPT, ID 
  FROM EMPLOYEE E 
  WHERE P.RESPEMP = E.EMPNO
)
WHERE P.RESPEMP > '000030' And Exists (
  SELECT * 
  FROM EMPLOYEE E 
  WHERE P.RESPEMP = E.EMPNO
);
```

**Comment by MaceWindu** [comment]
```cs
query.Update(table, r => new TableEntity { Field1 = ..., Field2 = ... })
```

E.g. https://github.com/linq2db/linq2db/blob/master/Tests/Linq/Update/UpdateTests.cs#L1418-L1428

**Reply by tb-mtg** [reply]
How do you do a delete with where exists?

DELETE FROM PROJECT P
WHERE P.RESPEMP > '000030' And Exists (
  SELECT * 
  FROM EMPLOYEE E 
  WHERE P.RESPEMP = E.EMPNO
);

**Reply by sdanyliv** [reply]
```cs
db.Projects
   .Where(p => p.Respemp > "000030")
   .Where(p => db.Employee.Any(e => p.Respemp == e.EmpNo))
   .Delete();
```

**Reply by tb-mtg** [reply]
But when I run: 
```
var list = db.GetTable<Project>()
  .Where(p => p.Respemp.CompareTo("000030")>0)
  .Where(p => db.GetTable<Employee>().Any(e => p.Respemp == e.EmpNo))
  .ToList();
```
I get SQL:
```
SELECT p.Respemp
FROM Project p
WHERE p.Respemp > '000030' AND (
    SELECT '.'
    FROM Employee e
    WHERE p.Respemp = e.EmpNo
    FETCH FIRST 1 ROWS ONLY
) IS NOT NULL
```

Which doesn't use **WHERE EXISTS** as expected and causes and error because of the **FETCH FIRST 1 ROWS ONLY** is unexpectedly added. This happend for all calls to **.Any()**.

The expected SQL should be:
```
SELECT p.Respemp
FROM Project p
WHERE p.Respemp > '000030' AND EXISTS (
    SELECT '.'
    FROM Employee e
    WHERE p.Respemp = e.EmpNo
) 
```
I have already raised this [issue](https://github.com/LinqToDB4iSeries/Linq2DB4iSeries/issues/81) but got no response.

How can I use **.Any()** to produce a valid SQL statement using **WHERE EXISTS**?

**Reply by sdanyliv** [reply]
You forgot to mention that this is the DB2 iSeries provider which is not maintained by linq2db team. It's unclear what the author was trying to address with this translation.

**Reply by tb-mtg** [reply]
Sorry about that but I wasn't getting any response from [LinqToDB4iSeries](https://github.com/LinqToDB4iSeries/Linq2DB4iSeries).

The best solution would be that given that **DB2iSeries** shares mostly the same SQL generated as **DB2**, it would be really helpful it could be added natively to [Linq2DB](https://github.com/linq2db/linq2db) as extra options to the enum DB2.DB2Version and (maybe **DB2.DB2Version.iSeries**). ,rather than relying on 3rd party [LinqToDB4iSeries](https://github.com/LinqToDB4iSeries/Linq2DB4iSeries).  

The benefits of this would be:
* First class support as I think the [LinqToDB4iSeries](https://github.com/LinqToDB4iSeries/Linq2DB4iSeries) library is inactive
* Removable of duplicated code

And before you can say it would be untestable, there is a site https://pub400.com that offers free access to the latest IBM i server (current on 7.5 version, where [LinqToDB4iSeries](https://github.com/LinqToDB4iSeries/Linq2DB4iSeries) currently only supports 7.5).

I would be very interested in helping where possible.

**Reply by MaceWindu** [reply]
We are concidering this option. I've already created account there some time ago, but didn't even had time to test it.
Also we will need a permission from provider developers to merge our codebases

**Comment by nsarris** [comment]
Hi @tb-mtg , @MaceWindu 

The project is not inactive, I just visit it from time to time to upgrade to latest linq2db and I was waiting for an RC to start working on it.

Pub400.com is really unstable, I had issues using it for the last few months untill I created new accounts and the team there is not responsive at all, at least whenever I tried. If you would like to merge the driver into linq2db I would be fine and could offer some assistance

For the matter at hand, the problem is that linq2db treats EXISTS as a function when in DB2i it is a PREDICATE meaning it is only used in WHERE clauses. @MaceWindu is there a way to differentiate somehow to allow EXISTS where it can be used?

**Reply by MaceWindu** [reply]
In v6 we fixed this inconsistency and replaced function with `SqlPredicate.Exists`

**Reply by nsarris** [reply]
Ok that's one part of it. Is there any suggestion on how to perform the convertion only outside the context of a WHERE clause?

**Comment by tb-mtg** [comment]
It would be really nice if DB2i could be merged into the main linq2db library with @nsarris assistance.
